RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,890 @@
+
+    let ofValue (a : 'a) : 'a Foo = Foo a
+
+    let apply ((Foo f) : ('a -> 'b) Foo) ((Foo a) : 'a Foo) : 'b Foo =
+        Foo (f a)
+
+
+type FooBuilder () =
+
+    member __.Yield (_ : unit) =
+        id
+
+    [<CustomOperation("apply")>]
+    member __.Apply (f : 'a Foo -> ('b -> 'c) Foo, foo : 'b Foo) : 'a Foo -> 'c Foo =
+        f >> (fun ff -> Foo.apply ff foo)
+
+    [<CustomOperation("into")>]
+    member __.Into (f : 'a Foo -> 'b Foo, a : 'a) : 'b Foo =
+        a |> Foo.ofValue |> f
+
+
+let foo = FooBuilder ()
+
+let test =
+    foo {
+        apply (Foo.ofValue 5)
+        apply (Foo.ofValue true)
+        apply (Foo.ofValue 12.34)
+        apply (Foo.ofValue "Hello")
+        into (fun i b f s -> if i > 3 && b then s else "Nope")
+    }
+```
+
+Thanks to [nickcowle](https://github.com/nickcowle) for providing this example.
+
+</p></details>
+
+# Compatibility
+[compatibility]: #compatibility
+
+## Is this a breaking change?
+
+This change should be backwards compatible.
+
+Existing computation expression builders with an `Apply` method should not change in behaviour, since usages of the builder would still need to add the new `let! ... and! ...` syntax to activate it. In particular, in the case of `let! ... return ...`, we will continue to only pick bind, as mentioned earlier.
+
+## What happens when previous versions of the F# compiler encounter this design addition as source code?
+
+Previous compiler versions reject the new `and!` and `anduse!` keywords:
+
+```
+error FS1141: Identifiers followed by '!' are reserved for future use
+```
+
+## What happens when previous versions of the F# compiler encounter this design addition in compiled binaries?
+
+Since the syntax is desugared into a method call on the builder object, after compilation usages of this feature will be usable with previous compiler versions.
+
+# Unresolved Questions
+[unresolved]: #unresolved-questions
+
+Is `anduse!` an acceptable keyword for when `and!` must also imply a call to `Using`? Are there sensible alternatives?
+
+There are various ways of desugaring `let! ... return ...` via one of `Map`, `Apply` or `Bind`. This RFC currently assumes that we are aiming for backward compatibility and hence doesn't consider desugaring to `Apply` at all. Is this the best choice? Alternatives include:
+
+* Using `Apply` in preference to `Bind` for `let! ... return` whenever `Apply` is defined (on the basis that that any reasonable `Apply` implementation will be functionally equivalent, but more efficient than the corresponding `Bind`)
+* Subsuming [the RFC for desugaring this to `Map`](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1048-ce-builder-map.md) and defining a hierarchy between `Map`, `Apply` and `Bind` and some attributes for those methods to allow the creators of builders to opt-in to "optimisations" that pick the least general (and hence hopefully most efficient) desugaring (i.e. pick `Map` if it is defined, else `Apply`, else `Bind`).
```

Comment:
I think it would be fantastic if this RFC could also address [FS 1048](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1048-ce-builder-map.md).  It does feel like the second choice here is the best one, i.e.  "pick `Map` if it is defined, else `Apply`, else `Bind`"

Reply:
Due to time constraints, I think I am going to have to kick the can down the road and let [FS 1048](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1048-ce-builder-map.md) handle this. I have sketched most of it out on top of the changes in this RFC and it looks fine to do (although I haven't given `use! ... return ...` all the attention it deserves).